src/Microsoft.AspNetCore.Blazor/Components/Provider.cs

```diff
@@ -85,6 +85,17 @@ public void SetParameters(ParameterCollection parameters)
             _renderHandle.Render(Render);
         }
 
+        internal override bool CanSupplyValue(Type requestedType, string requestedName)
+        {
+            if (!requestedType.IsAssignableFrom(typeof(T)))
+            {
+                return false;
+            }
+
+            return (requestedName == null && Name == null) // Match on type alone
+                || string.Equals(requestedName, Name, StringComparison.Ordinal); // Also match on name
+        }
```

Comment:
So really how we should think about `name` based on this code is that a named value cannot be bound to an unnamed reciever. I think it makes sense, I just hadn't gone through it mentally yet.

Reply:
Yes. I'm open to changing the design, but the idea was simply that if you choose to use "named slots", you're opting into that extra strictness on both the provider and consumer side.

My reasoning is that you'd only use the named slots if the value you're supplying is of a common type that might reasonably clash with other uses (e.g., `string`, `int`, `IPrincipal`, etc.). In that case you absolutely don't want your value showing up randomly in unrelated places just because they are asking for a `string`, so it's safer to force the consumer to declare the same named slot.

If the type you're supplying is unique enough (e.g., it's a special type you've defined purely for the purpose of cascading info to descendants), then you would have no reason for declaring a name too.

If it turns out I'm misjudging the use cases then we can certainly tweak the rules.